File: Golang/BinaryTreeInorderTraversal/BinaryTreeInorderTraversal.go
package BinaryTreeInorderTraversal

//Given a binary tree, return the inorder traversal of its nodes' values.
//
//For example:
//Given binary tree [1,null,2,3],
//1
//\
//2
///
//3
//return [1,3,2].
//
//Note: Recursive solution is trivial, could you do it iteratively?

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//Accepted.
func inorderTraversal(root *TreeNode) []int {
	list := make([]int, 0)
	if root == nil {
		return list
	}

	stack := make([]*TreeNode, 0)
	for ; root != nil || len(stack) != 0; {
		for ; root != nil; {
			stack = append(stack, root)
			root = root.Left
		}

		root = stack[len(stack)-1]
		list = append(list, root.Val)
		stack = removeIndex(stack, len(stack)-1)
		root = root.Right
	}

	return list
}

func removeIndex(s []*TreeNode, index int) []*TreeNode {
	return append(s[:index], s[index+1:]...)
}

//Recursive solution.
//Accepted.
//func inorderTraversal(root *TreeNode) []int {
//	list := make([]int, 0)
//
//	if root == nil {
//		return list
//	}
//	if root.Left != nil {
//		for _, i := range inorderTraversal(root.Left) {
//			list = append(list, i)
//		}
//	}
//
//	list = append(list, root.Val)
//
//	if root.Right != nil {
//		for _, i := range inorderTraversal(root.Right) {
//			list = append(list, i)
//		}
//	}
//	return list
//}

File: Swift/LeetCode/LeetCode/BinaryTreeInorderTraversal.swift
//
//  BinaryTreeInorderTraversal.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 28/11/2017.
//  Copyright © 2017 lizhaotailang. All rights reserved.
//
//  Given a binary tree, return the inorder traversal of its nodes' values.
//
//  For example:
//  Given binary tree [1,null,2,3],
//  1
//  \
//  2
//  /
//  3
//  return [1,3,2].
//
//  Note: Recursive solution is trivial, could you do it iteratively?
//
//  Accepted. See [BinaryTreeInorderTraversalTests](./LeetCodeTests/BinaryTreeInorderTraversalTests.swift) for test cases.
//

import Foundation

class BinaryTreeInorderTraversal {
    
//    Recursive solution.
//    func inorderTraversal(_ root: TreeNode?) -> [Int] {
//        var list: [Int] = []
//        if root == nil {
//            return list
//        }
//
//        if root?.left != nil {
//            inorderTraversal(root?.left).forEach{ list.append($0) }
//        }
//
//        list.append((root?.val)!)
//
//        if root?.right != nil {
//            inorderTraversal(root?.right).forEach{ list.append($0) }
//        }
//
//        return list
//    }
    
    // Iterative solution.
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        var list: [Int] = []
        if root == nil {
            return list
        }
        
        var node = root
        var stack: [TreeNode?] = []
        while node != nil || !stack.isEmpty {
            while node != nil {
                stack.append(node)
                node = node?.left
            }
            
            node = stack[stack.count - 1]
            stack.remove(at: (stack.count - 1))
            list.append((node?.val)!)
            node = node?.right
        }
        
        return list
    }
    
    public class TreeNode: Equatable {
        
        public var val: Int
        public var left: TreeNode?
        public var right: TreeNode?
        
        public init(_ val: Int) {
            self.val = val
            self.left = nil
            self.right = nil
        }
        
        public static func ==(lhs: BinaryTreeInorderTraversal.TreeNode, rhs: BinaryTreeInorderTraversal.TreeNode) -> Bool {
            return lhs.val == rhs.val && lhs.left == rhs.left && lhs.right == rhs.right
        }
        
    }
    
}
